### PR TITLE
#3057. Add more try-finally tests

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: TypeSystem/flow-analysis/reachability_try_finally_A02_t05.dart [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+/// `try B1 finally B2` then:
+/// - Let `before(B1) = split(before(N))`
+/// - Let `before(B2) = split(join(drop(after(B1)),
+///     conservativeJoin(before(N), assignedIn(B1), capturedIn(B1))))`
+/// - Let `after(N) = restrict(after(B1), after(B2), assignedIn(B2))`
+///
+/// @description Checks that `before(B2) = split(join(drop(after(B1)),
+/// conservativeJoin(before(N), assignedIn(B1), capturedIn(B1))))`. Test that if
+/// some variable is assigned in a `catch` part of `B1` then it is
+/// "possibly assigned" in `B2`.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply TypeSystem/flow-analysis/reachability_try_finally_A02_t05.dart:15]: I created https://github.com/dart-lang/language/issues/4327 in order to clarify how to deal with try/catch/finally.

File: TypeSystem/flow-analysis/reachability_try_finally_A02_t05.dart [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+test1() {
+  late int i;
+  try {
+    print("To avoid empty body");
+  } catch (_) {
+    foo();
+    i = 42;
+  } finally {
+    i; // Possibly assigned
```
[the remark this answers is not in the document]
Comment (eernstg) [reply TypeSystem/flow-analysis/reachability_try_finally_A02_t05.dart:33]: With this, I'd say that `i = 42` is unreachable code and there is no conservative join, so `i` is definitely not assigned when we reach `finally`.

I can see that the analyzer and CFE think otherwise, though. I don't really understand why there would be a need for a conservative join involving each of the `catch` blocks, but this seems to be what the implementation is doing. I created https://github.com/dart-lang/language/issues/4327 in order to clarify the situation.

[the remark this answers is not in the document]
Comment (eernstg) [reply TypeSystem/flow-analysis/reachability_try_finally_A02_t05.dart:33]: This means that the outcome of the following cases is similarly in need of a clarification.

File: TypeSystem/flow-analysis/reachability_try_finally_A02_t06.dart [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+/// `try B1 finally B2` then:
+/// - Let `before(B1) = split(before(N))`
+/// - Let `before(B2) = split(join(drop(after(B1)),
+///     conservativeJoin(before(N), assignedIn(B1), capturedIn(B1))))`
+/// - Let `after(N) = restrict(after(B1), after(B2), assignedIn(B2))`
+///
+/// @description Checks that `before(B2) = split(join(drop(after(B1)),
+/// conservativeJoin(before(N), assignedIn(B1), capturedIn(B1))))`. Test that if
+/// some variable is assigned in a `catch` part of `B1` then it is
+/// "possibly assigned" in `B2`.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply TypeSystem/flow-analysis/reachability_try_finally_A02_t06.dart:15]: This would apply for all the following test libraries as well.

**Review by eernstg — COMMENTED** [review]
Here's a comment from Paul about `finally` blocks that we should keep in mind: https://github.com/dart-lang/sdk/issues/60503#issuecomment-2790573846. See also https://github.com/dart-lang/language/issues/4327 that I created in order clarify the situation.

I think it would be useful to block several of these tests for a little while because they depend in such a detailed manner on the flow analysis of try/catch/finally.

Perhaps the most convenient approach would be to make them part of a separate PR.

File: TypeSystem/flow-analysis/reachability_try_finally_A02_t05.dart [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  int v;
+  C(this.v);
+}
+
+Never foo() => throw "Never";
+
+test1() {
+  late int i;
+  try {
+    print("To avoid empty body");
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply TypeSystem/flow-analysis/reachability_try_finally_A02_t05.dart:28]: Here `i` is initialized in a `catch` block with is guaranteed to be dead code in case of an empty `try` block. To avoid side effects I made a body of `try` a non-empty. But probably better approach is not to do this but add a separate test checking empty body of `try`. I'll update this.

**Comment by sgrekhov** [comment]
Blocked by https://github.com/dart-lang/language/issues/4327
